### ISSUE TITLE
State Transfer Errors Running a **busywork** target with multiple clients

### ISSUE THREAD

**Issue by bcbrock** [hyperledger-archives/fabric#1857]
## Description

@jyellick Jason, When running busywork PBFT Batch with 2 clients, one of the peers seems to get stuck in state transfer. In this example it is VP2.

[logs.tar.gz](https://github.com/hyperledger/fabric/files/317403/logs.tar.gz)
## Describe How to Reproduce

This is the target I'm trying to run, it does not make it past 2 clients. It fails even w/o the PBFT configs of the network shown below. One client tests seem to work OK.

If you run this you will see a "...state may be inconsistent ..." error. However if you "curl" the "/chain" API of the peer that gives this error you will see that it is not making progress. It is stuck at a block height less than the expected 16387 shown by the other peers.

```.PHONY: s2b 
s2b:
    for clients in 2 4 8 16 32; do \
    CORE_PBFT_GENERAL_TIMEOUT_REQUEST=5s \
    CORE_PBFT_GENERAL_TIMEOUT_VIEWCHANGE=5s \
    CORE_PBFT_GENERAL_VIEWCHANGEPERIOD=0 \
    CORE_PBFT_GENERAL_NULLREQUEST=1s \
    $(NETWORK) -batch 4; \
    set -e; ./driver \
        -clients $$clients \
        -transactions $$((16384 / $$clients)) \
        -peerBurst 100 \
        -interlock \
        -keepLog \
        -checkAgreement -dupCheck; \
    done

```
```

**Comment by bcbrock** [comment]
It is not necessary to use multiple clients, it seems to be sufficient for the network to be receiving transactions on multiple peers simultaneously. If in the above you run with 1 client and change 

```
-peerBurst 100
```

 to 

```
-peerBurst 10 -netBurst 10
```

 it will fail the same way. (`-peerBurst 100` sends 100 TX to a random peer then waits for them to all complete before moving on. `-peerBurst 10 -netBurst 10` sends 10 bursts of 10 TX to random peers then waits.)

UPDATE: This is bizarre - The multi-client runs started working this morning. But the -peerBurst/-netBurst runs consistently fail with 1 client.

ANOTHER UPDATE: There must be a subtle timing dependency here. I noticed that the "driver" processes were taking a lot of CPU time. I assumed this was happening when they polled for transaction completion, so added some code to try to delay the polling loops. This somehow makes the runs work. 

The bottom line is that it may be difficult to reproduce this issue.

**Comment by jyellick** [comment]
@bcbrock I think I have a set of logs locally which exhibits this.  I'll be investigating today.

**Comment by bcbrock** [comment]
Thanks for looking into this Jason.

I have also just observed a different issue. I have only seen this once. In this case, all of the peers report the same block height, and in fact their blockchains are identical (as reported by the checkAgreement script). However the state of one of the peers is not complete. This peer is missing about 100 "counts". The chaincode reports that it executed 16274 (of 16328) TX, and the state of its counter is also 16274. So its blockchain contains all TX, but not its state. I've repeated the query long after the run completed with no change in the result. I did not have debugging logs on during this run, but these are the last few message captured on the failing peer:

```
[33m08:16:36.938 [consensus/obcpbft] recvCommit -> WARN 0cc[0m Replica 2 ignoring commit for view=0/seqNo=1728: not in-wv, in view 0, high water mark 1670
[33m08:16:36.938 [consensus/obcpbft] recvCommit -> WARN 0cd[0m Replica 2 ignoring commit for view=0/seqNo=1725: not in-wv, in view 0, high water mark 1670
[33m08:16:36.938 [consensus/obcpbft] recvCommit -> WARN 0ce[0m Replica 2 ignoring commit for view=0/seqNo=1724: not in-wv, in view 0, high water mark 1670
[33m08:16:36.938 [consensus/obcpbft] recvCommit -> WARN 0cf[0m Replica 2 ignoring commit for view=0/seqNo=1725: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d0[0m Replica 2 ignoring commit for view=0/seqNo=1726: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d1[0m Replica 2 ignoring commit for view=0/seqNo=1729: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvPrepare -> WARN 0d2[0m Replica 2 ignoring prepare for view=0/seqNo=1729: not in-wv, in view 0, low water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d3[0m Replica 2 ignoring commit for view=0/seqNo=1727: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvPrepare -> WARN 0d4[0m Replica 2 ignoring prepare for view=0/seqNo=1730: not in-wv, in view 0, low water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d5[0m Replica 2 ignoring commit for view=0/seqNo=1726: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d6[0m Replica 2 ignoring commit for view=0/seqNo=1730: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d7[0m Replica 2 ignoring commit for view=0/seqNo=1728: not in-wv, in view 0, high water mark 1670
[33m08:16:36.939 [consensus/obcpbft] recvCommit -> WARN 0d8[0m Replica 2 ignoring commit for view=0/seqNo=1727: not in-wv, in view 0, high water mark 1670
[33m08:16:36.940 [consensus/obcpbft] recvCommit -> WARN 0d9[0m Replica 2 ignoring commit for view=0/seqNo=1729: not in-wv, in view 0, high water mark 1670
[33m08:16:36.940 [consensus/obcpbft] recvCommit -> WARN 0da[0m Replica 2 ignoring commit for view=0/seqNo=1730: not in-wv, in view 0, high water mark 1670
[33m08:16:36.940 [consensus/obcpbft] recvCommit -> WARN 0db[0m Replica 2 ignoring commit for view=0/seqNo=1728: not in-wv, in view 0, high water mark 1670
[33m08:16:36.941 [consensus/obcpbft] recvCommit -> WARN 0dc[0m Replica 2 ignoring commit for view=0/seqNo=1729: not in-wv, in view 0, high water mark 1670
[33m08:16:36.941 [consensus/obcpbft] recvCommit -> WARN 0dd[0m Replica 2 ignoring commit for view=0/seqNo=1730: not in-wv, in view 0, high water mark 1670
[33m08:16:36.941 [consensus/obcpbft] weakCheckpointSetOutOfRange -> WARN 0de[0m Replica 2 is out of date, f+1 nodes agree checkpoint with seqNo 1720 exists but our high water mark is 1710
[33m08:16:37.063 [consensus/obcpbft] ProcessEvent -> WARN 0df[0m Replica 2 recovered to seqNo 1730 but our low watermark has moved to 1760
```

```
```

**Comment by bcbrock** [comment]
@jyellick Jason, I am not seeing this reliably fixed, even with a single client. I am seeing a little bit of everything: Working perfectly without issues, state transfer happening at the end of TX invocation that finishes within a couple of seconds and then checks perfectly, and cases where 1 peer appears to get stuck early in the process and never recover.

**Comment by jyellick** [comment]
@bcbrock Is this stress2b?  Do you have debug logs or a way I can reproduce locally?

Edit: Also, be aware that transaction duplication is still possible under stress induced network latencies, but a 'correct' solution to this is non-trivial, so we are designing this now.

**Comment by bcbrock** [comment]
@Jyellick, yes this is stress2b, with a 10-second delay between network startup and the driver. I was able to reproduce with debug logs. Peer 1 is stuck in this run at block height 1900 of 5003. The logs are too big for Github. I put them at

```
/gsa/ausgsa/home/b/c/bcbrock/web/public/1857.tgz
```

There is no evidence of queue overflow in this run, but lots of warnings of various kinds.

**Comment by jyellick** [comment]
@bcbrock Thanks for the logs, I see exactly what is going on here.

vp1 is too far behind to recover via state deltas, so it requests a state snapshot from someone on the network.  The state it gets is for block 5002.  The last checkpoint it knows about is block 5000, so, vp1 cannot validate any blocks past 5000, and therefore cannot validate its state.  It is waiting to hear about a checkpoint for a higher block, but because the network is no longer processing new transactions, one is not forthcoming.  

You might try seeing `CORE_PBFT_GENERAL_TIMEOUT_NULLREQUEST=3s` to enable the periodic null requests.  This will cause the network to emit checkpoints periodically, even though you are not submitting new transactions.

**Comment by bcbrock** [comment]
@jyellick Why would we not add `CORE_PBFT_GENERAL_TIMEOUT_NULLREQUEST=3s` (or some other value) to the YAML file as a default?

**Comment by jyellick** [comment]
@bcbrock Enabling null requests by default has been discussed with @kchristidis @tuand27613 @corecode and others, and we think it would be a sane thing to do.

There are two reasons we have not done this.
1. All of our current tests were written under the assumption that null requests are not enabled.  Enabling null requests by default might break some tests, and hide breakage in others, so an audit of all the PBFT tests would be required.
2. Most all of the non-CI type testing that has been done, has been done with the null requests disabled, so we're correspondingly less confident in this as a default.

Ultimately, I absolutely agree, this should be a default, the question is when and how.

**Comment by bcbrock** [comment]
@jyellick, These are logs from a stress2b test modified as you suggested. I'm proposing this as the temporary form of `stress2b`. 

```
stress2b: 
    @CORE_PBFT_GENERAL_TIMEOUT_NULLREQUEST=3s \
    $(NETWORK) -batch 4
    @echo "Sleeping 10 seconds to allow peers to interlock"; sleep 10
    @$(STRESS2)
```

Peer 3 gets stuck here at block height 57. Unfortunately I do not have debug logs (It takes so much time to write them that it alters the behavior).

[1857a.tar.gz](https://github.com/hyperledger/fabric/files/326708/1857a.tar.gz)

**Comment by jyellick** [comment]
@bcbrock I think that's a sane definition for stress2b.  I have been running without DEBUG on with stress2b as I too noticed some issues that didn't seem to appear when DEBUG was on, but it runs reliably for me locally now.  Can you tell me any more about your environment so that maybe I can try executing on something similar, then manually add in raise the severity of certain log statements to hopefully capture what is going on without affecting the performance.

**Comment by bcbrock** [comment]
@jyellick I observed this last failure running on a 20-core POWER8 server. The busywork test is all that's running on the machine. I have tuned the system for performance: 1) Export GOGC=400 to reduce garbage collection overhead; 2) I have modified Golang 1.6 as described here: 

https://github.com/hyperledger/fabric/wiki/Go-Performance-Portal#morestack

to increase the default stack size. This is a rare failure, usually the runs either work or fail the duplicate test. If you can pull in my updates to busywork from PR #1946 a lot of the reported errors I assume you are seeing should go away. 

Right now I am trying the idea of increasing the number of TX from 10K to 100K to see if that makes the error more likely. I'll let you know.

**Comment by jyellick** [comment]
@bcbrock Using the latest busywork code from #1946 and no debug, I think I may be able to reproduce this.  Even long after the tests, I am still seeing the shortened chain, which I would expect to have corrected itself.  I will be investigating.

**Comment by bcbrock** [comment]
@jyellick Thanks for continuing to look into this. I can still surface various errors under heavy stress, e.g., using 32 clients. Here are debug logs from the latest fail. The symptom is that a client has sent a batch of 100 TX, but it never sees confirmation within 5 minutes. Note that Peer 2 has a lot more to say (based on log file size) than the other peers.

```
/gas/ausgsa/home/b/c/bcbrock/web/public/logs.tar.gz
```

This is the busywork/counters make target I am running now:

```
.PHONY: s2b 
s2b:
    for clients in 32 32 32 32 32; do \
    CORE_PBFT_GENERAL_TIMEOUT_NULLREQUEST=3s \
    CORE_PEER_VALIDATOR_CONSENSUS_BUFFERSIZE=100000 \
    $(NETWORK) -peerLogging debug -batch 4; \
    sleep 10; \
    set -e; ./driver \
        -clients $$clients \
        -transactions $$((16384 / $$clients)) \
        -peerBurst 10 -netBurst 10 \
        -interlock \
        -keepLog \
        -interlockTimeout 5m \
        -finalWait 1m \
        -checkAgreement ""; \
    done
```

**Comment by jyellick** [comment]
This has been reported by another user and should be a candidate for inclusion in 0.5.

**Comment by kchristidis** [comment]
Is the related to the checkpoint garbage collection issue?

**Comment by jyellick** [comment]
Yes, this is actually a combination of two bugs.  One half is the checkpoint garbage collection, the other half is... something else, still trying to track down exactly how it's happening.  I'd like to hold off fixing the garbage collection issue until the second bug is tracked down, as the fix would likely mask the problem.

**Comment by jyellick** [comment]
@kchristidis @bcbrock The second issue at work turned out to be non-determinism in the busywork counters chaincode.  I've fixed it in the PR referenced above, and no longer see busywork stress2b failures.

The pbft core was still mishandling the checkpoint garbage collection, so I will be submitting a second PR for this issue later today.

**Comment by jyellick** [comment]
The PRs above (#1969 for busywork and #1973 for checkpoint garbage collection in master) fix the original reported 1857 issue for me.

PR #1971 fixes the checkpoint garbage collection in 0.5.  If busywork should be patched in 0.5 as well, the patch is trivial and I can submit an additional PR for this.
